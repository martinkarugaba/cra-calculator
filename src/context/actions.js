const ACTIONS = {
  HANDLE_CLICK: 'HANDLE_CLICK',
  FETCH_SUCCESS: 'FETCH_SUCCESS',
  FETCH_ERROR: 'FETCH_ERROR',
};

export default ACTIONS;
