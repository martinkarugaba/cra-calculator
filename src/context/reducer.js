import ACTIONS from './actions';
import calculate from '../logic/calculate';

const reducer = (state, action) => {
  switch (action.type) {
    case ACTIONS.HANDLE_CLICK:
      return {
        ...state,
        result: calculate(state.result, action.payload),
      };
    case ACTIONS.FETCH_SUCCESS:
      return {
        ...state,
        data: action.payload,
        loading: false,
        error: null,
      };
    case ACTIONS.FETCH_ERROR:
      return {
        ...state,
        data: null,
        loading: false,
        error: true,
      };
    default:
      return state;
  }
};

export default reducer;
