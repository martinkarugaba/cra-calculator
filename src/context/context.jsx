import {
  createContext,
  useContext,
  useReducer,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';
import reducer from './reducer';
import ACTIONS from './actions';

const defaultState = {
  data: null,
  loading: true,
  error: false,
  result: { total: null, next: null, operation: null },
};

const GlobalContext = createContext();
export const useGlobalContext = () => useContext(GlobalContext);

const AppContext = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, defaultState);

  const handleClick = (buttonName) => {
    dispatch({ type: ACTIONS.HANDLE_CLICK, payload: buttonName });
  };

  const fetchData = async () => {
    try {
      const url = 'https://api.api-ninjas.com/v1/quotes?category=humor';
      const apiKey = process.env.REACT_APP_QUOTES_API_KEY;
      const response = await fetch(url, {
        headers: {
          'Content-Type': 'application/json',
          'X-Api-Key': apiKey,
        },
      });
      const data = await response.json();
      dispatch({ type: ACTIONS.FETCH_SUCCESS, payload: data[0] });
    } catch (error) {
      dispatch({ type: ACTIONS.FETCH_ERROR, payload: error.message });
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  return (
    <GlobalContext.Provider
      value={{
        ...state,
        handleClick,
      }}
    >
      {children}
    </GlobalContext.Provider>
  );
};

AppContext.propTypes = {
  children: PropTypes.node.isRequired,
};

export default AppContext;
