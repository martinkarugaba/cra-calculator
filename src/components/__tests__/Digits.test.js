import React from "react";
import { render } from "@testing-library/react";
import SingleButton from "../SingleButton";
import { useGlobalContext } from "../../context/context";

jest.mock("../../context/context", () => ({
  __esModule: true,
  useGlobalContext: jest.fn(),
}));

describe("SingleButton component", () => {
  test("renders correctly", () => {
    useGlobalContext.mockReturnValue({
      handleClick: jest.fn(),
    });

    const props = {
      buttonName: "Test Button",
      basis: "basis-class",
      bg: "bg-class",
    };

    const { container } = render(<SingleButton {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
