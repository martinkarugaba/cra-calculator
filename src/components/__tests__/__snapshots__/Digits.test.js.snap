// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleButton component renders correctly 1`] = `
<button
  class="undefined flex justify-center items-center undefined hover:bg-slate-400 text-black h-[60px] border-r border-t text-xs cursor-pointer"
  type="button"
/>
`;
